components/sup/src/manager/file_watcher.rs

```diff
@@ -26,3439 +31,1461 @@ use std::{collections::{hash_map::Entry,
+        let config = Config::default().with_poll_interval(Duration::from_millis(WATCHER_DELAY_MS));
+        let watcher = RefCell::new(W::new(tx, config).map_err(Error::NotifyCreateError)?);
+        let p = path.into();
+        let watched_file = WatchedFile::new(&p, !ignore_initial)?;
+        let active_watches = RefCell::new(HashSet::new());
+        let mut file_watcher = Self { callbacks,
+                                      watcher,
+                                      rx,
+                                      watched_file,
+                                      active_watches };
+        file_watcher.manage_watches()?;
+        Ok(file_watcher)
+    }
+
+    /// manage_watches() wraps up the essential functions around "the watch" for the WatchedFile
+    /// being watched by the FileWatcher.  Its really meant to be called in two places.  First,
+    /// during construction of FileWatcher instances.  Second, as late as possible in the event
+    /// handling process. Its called during construction because we need to assess "the state of
+    /// the world" when the watch of the file begins so that we know how to respond to the next
+    /// event received. We then defer calling this again until as late as possible in the process
+    /// of responding to an event because our response needs to be within the context of the last
+    /// action we took.
+    fn manage_watches(&mut self) -> Result<()> {
+        trace!("FileWatcher::manage_watches()");
+        self.clear_active_watches()?;
+        self.watched_file.state = self.watched_file.assess_state();
+        self.set_new_watches()
+    }
+
+    fn set_new_watches(&self) -> Result<()> {
+        trace!("FileWatcher::set_new_watches()");
+        match &self.watched_file.state {
+            WatchedFileState::ExistentFile => {
+                Ok(self.watch(&self.watched_file.path)
+                       .and(self.watch(&self.watched_file.directory))?)
+            }
+            WatchedFileState::ExistentDirectory => Ok(self.watch(&self.watched_file.directory)?),
+            WatchedFileState::NonExistent => Ok(()),
+            WatchedFileState::Invalid(_) => {
+                let s = self.watched_file
+                            .path
+                            .as_os_str()
+                            .to_string_lossy()
+                            .to_string();
+                Err(Error::FileNotFound(s))
+            }
+        }
+    }
+
+    fn watch(&self, path: &Path) -> Result<()> {
+        trace!("FileWatcher::watch({:?})", path);
+        self.active_watches.borrow_mut().insert(path.to_path_buf());
+        self.watcher
+            .borrow_mut()
+            .watch(path, RecursiveMode::NonRecursive)
+            .map_err(Error::NotifyError)
+    }
+
+    fn clear_active_watches(&self) -> Result<()> {
+        trace!("FileWatcher::clear_active_watches()");
+        let mut laws = self.active_watches.borrow_mut();
```

Comment:
Would it make more sense to create a new Watcher here which should effectively clear out the old watches? Maybe then you don't need to maintain a separate active_watches hashset?